File: apina-core/src/main/java/fi/evident/apina/java/reader/JavaModelLoader.java
package fi.evident.apina.java.reader;

import fi.evident.apina.java.model.JavaClass;
import fi.evident.apina.java.model.JavaModel;
import fi.evident.apina.java.model.type.JavaType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.LinkedHashSet;
import java.util.Set;

public final class JavaModelLoader {

    private JavaModelLoader() { }

    private static final Logger log = LoggerFactory.getLogger(JavaModelLoader.class);

    public static JavaModel load(Classpath classpath) throws IOException {
        JavaModel classes = new JavaModel();

        Set<JavaType> duplicates = new LinkedHashSet<>();

        ClassPathScanner.processAllClasses(classpath, in -> {
            JavaClass aClass = ClassMetadataReader.loadMetadata(in);
            if (!classes.containsClass(aClass.getName())) {
                classes.addClass(aClass);
            } else {
                duplicates.add(aClass.getType());
            }
        });

        if (!duplicates.isEmpty()) {
            log.warn("There were {} classes with multiple definitions in classpath. Ignoring duplicate definitions.", duplicates.size());
            log.debug("Classes with multiple definitions: {}", duplicates);
        }

        return classes;
    }
}

File: apina-core/src/main/kotlin/fi/evident/apina/java/reader/JavaModelLoader.kt
package fi.evident.apina.java.reader

import fi.evident.apina.java.model.JavaModel
import fi.evident.apina.java.model.type.JavaType
import org.slf4j.LoggerFactory
import java.util.*
import kotlin.system.measureTimeMillis

private object JavaModelLoader
private val log = LoggerFactory.getLogger(JavaModelLoader::class.java)

fun Classpath.loadModel(): JavaModel {
    val classes = JavaModel()

    val duplicates = LinkedHashSet<JavaType>()

    val processingMillis = measureTimeMillis {
        processAllClasses { inputStream ->
            val aClass = ClassMetadataReader.loadMetadata(inputStream)
            if (!classes.containsClass(aClass.name)) {
                classes.addClass(aClass)
            } else {
                duplicates += aClass.type
            }
        }
    }

    log.debug("Loaded {} classes in {} ms", classes.classes.size, processingMillis)

    if (duplicates.any()) {
        log.warn("There were {} classes with multiple definitions in classpath. Ignoring duplicate definitions.", duplicates.size)
        log.debug("Classes with multiple definitions: {}", duplicates)
    }

    return classes
}
